packages/cozy-realtime/src/CozyRealtime.js

```diff
@@ -1,360 +1,534 @@
+    this.bindEventHandlers()
   }
 
-  let id, handler
-  if (!handlerOrUndefined) {
-    id = undefined
-    handler = handlerOrId
-  } else {
-    id = handlerOrId
-    handler = handlerOrUndefined
+  /**
+   * Start all handlers and try to connect if online
+   * @private
+   */
+  start() {
+    if (!this.isStarted) {
+      logger.info('started')
+      this.isStarted = true
+      this.retryManager.reset()
+      this.subscribeGlobalEvents()
+      this.subscribeClientEvents()
+      this.emit('start')
+      if (isOnline()) this.connect()
+    }
   }
 
-  return { id, handler }
-}
-
-const VALID_EVENTS = ['created', 'updated', 'deleted', 'notified']
-
-const validateParameters = (eventName, type, id, handler) => {
-  let msg
-
-  if (!VALID_EVENTS.includes(eventName)) {
-    msg = `'${eventName}' is not a valid event, valid events are 'created', 'updated' or 'deleted'.`
+  /**
+   * Create and attach a WebSocket to the current instance
+   *
+   * @private
+   * @returns {WebSocket} the created websocket
+   */
+  createSocket() {
+    if (this.websocket) {
+      logger.warn(
+        'Unattended replacement of an existing socket, this should not happen'
+      )
+      this.rejectWebSocket(this.websocket)
+    }
+    logger.info('creating a new websocket…')
+    const url = getUrl(this.client)
+    try {
+      this.websocket = new WebSocket(url, doctype)
+      this.websocket.authenticated = false
+      this.websocket.onmessage = this.onWebSocketMessage
+      this.websocket.onerror = this.onWebSocketError
+      this.websocket.onopen = this.onWebSocketOpen
+      this.websocket.onclose = this.onWebSocketClose
+      return this.websocket
+    } catch (error) {
+      this.onWebSocketError(error)
+    }
   }
 
-  if (typeof type !== 'string') {
-    msg = `'${type}' is not a valid type, it should be a string.`
+  /**
+   * Connect a new websocket as soon as possible
+   * @private
+   */
+  async connect() {
+    logger.info('connecting…')
+    await this.retryManager.waitBeforeNextAttempt()
+    this.createSocket()
   }
 
-  if (id && eventName === 'created') {
-    msg = `The 'id' should not be specified for 'created' event.`
+  /**
+   * Throw the previous socket and connect a new one
+   * @private
+   */
+  async reconnect() {
+    if (this.hasWebSocket()) this.rejectWebSocket()
+    this.connect()
   }
 
-  if (typeof handler !== 'function') {
-    msg = `The handler '${handler}' should be a function.`
+  /**
+   * Remove all handlers on a websocket to avoid callbacks from an old rejected socket
+   * @private
+   */
+  rejectWebSocket() {
+    this.emit('disconnected')
+    if (this.hasWebSocket()) {
+      logger.info('trashing the previous websocket…')
+      this.websocket.onmessage = null
+      this.websocket.onerror = null
+      this.websocket.onopen = null
+      this.websocket.onclose = null
+      try {
+        this.websocket.close()
+      } catch (e) {
+        // void
+      }
+      this.websocket = null
```

Comment:
If close returns an error,it will be catched line 109. As the catch blockdoes not raise itself, the process will continue  and line 112 will be executed, setting the attribute  to null. Am I  wrong ?

That said, a finally block is indeed clearer, I will do the change.